File: Suldin_Vyacheslav/FourthTerm/Block2/Task4/Task4/ThreadPool.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Task4
{
    public class ThreadPool : IDisposable
    {
        private const int threadCount = 5;
        public Queue<Action> actions { get; private set; }
        public List<Thread> threads { get; private set; }
        private volatile bool stop = false;

        public ThreadPool()
        {
            actions = new Queue<Action>();
            threads = new List<Thread>();
            for (var i = 0; i < threadCount; i++)
            {
                threads.Add(new Thread(Operate));
            }
            for (var i = 0; i < threadCount; i++)
            {
                threads[i].Start();
            }
        }

        public void Enqueue(Action action)
        {
            if (stop)
            {
                throw new InvalidOperationException("Enqueue in disposed thread pool");
            }
            lock (actions)
            {
                actions.Enqueue(action);
                Monitor.Pulse(actions);
            }
        }

        private void Operate()
        {
            while (true)
            {
                Action action = null;
                lock (actions)
                {
                    if (!actions.Any() && stop)
                    {
                        break;
                    }
                    while (!actions.Any() && !stop)
                    {
                        Monitor.Wait(actions);
                    }
                    if (actions.Any())
                    {
                        action = actions.Dequeue();
                    }
                }

                if (action != null)
                {
                    try
                    {
                        action();
                    }
                    catch (Exception ex)
                    {
                        Console.WriteLine($"exeption [{ex}]");
                    }
                }
            }
        }

        public void Dispose()
        {
            stop = true;

            lock (actions)
            {
                Monitor.PulseAll(actions);
            }

            for (var i = 0; i < threadCount; i++)
            {
                threads[i].Join();
            }

            threads.Clear();
        }
    }
}

File: VyacheslavBuchin/Term4/Block2/ThreadPool/ThreadPool/ThreadPool.fs
namespace ThreadPool

open System
open System.Collections.Generic
open System.Threading
open System.Linq

type ThreadPool(threadCount, initTasks : (unit -> unit) seq) =
    [<VolatileField>]
    let mutable isDisposed = false
    let tasks = Queue<unit -> unit>(initTasks)
    let dequeueSync = obj()

    let getTask () =
        let haveAnyTask, task = tasks.TryDequeue()
        if haveAnyTask then Some task
        else None

    let initThread _ =
        let threadWork () =
            let mutable isRunning = true
            while isRunning do
                lock dequeueSync (fun () ->
                        if not isDisposed && (tasks.Any() |> not) then
                            Monitor.Wait dequeueSync |> ignore
                        if isDisposed && (tasks.Any() |> not) then
                            isRunning <- false
                            None
                        else getTask()
                    )
                |> Option.iter (fun f -> f())
        let thread = Thread(threadWork)
        thread.Start()
        thread

    let threads = Array.init threadCount initThread

    new(threadCount) =
         new ThreadPool(threadCount, Seq.empty)

    member x.Enqueue task =
        let enqueue () =
            tasks.Enqueue task
            Monitor.Pulse dequeueSync
        if isDisposed then raise (InvalidOperationException())
        lock dequeueSync enqueue

    interface IDisposable with
        member x.Dispose() =
            isDisposed <- true
            lock dequeueSync (fun () -> Monitor.PulseAll dequeueSync)
            for thread in threads do thread.Join()
